mobile_verifier/src/reward_shares/mod.rs

```diff
@@ -419,6 +399,66 @@ impl ServiceProviderShares {
     }
 }
 
+pub struct ServiceProviderRewards {
+    pub rewards: HashMap<i32, ServiceProviderReward>,
+}
+
+pub struct ServiceProviderReward {
+    pub for_service_provider: Decimal,
+    pub for_promotions: Decimal,
+}
+
+impl ServiceProviderRewards {
+    pub fn get_total_rewards(&self) -> Decimal {
+        self.rewards
+            .values()
+            .map(|x| x.for_service_provider + x.for_promotions)
+            .sum()
+    }
+
+    pub fn get_total_rewards_allocated_for_promotion(&self) -> Decimal {
+        self.rewards.values().map(|x| x.for_promotions).sum()
+    }
+
+    /// Take the rewards allocated for promotion from a service provider, leaving none
+    /// left. If any rewards allocated for promotion are left by the time we call
+    /// into_mobile_reward_share, they will be converted to service provider rewards.
+    pub fn take_rewards_allocated_for_promotion(&mut self, sp: &i32) -> Decimal {
+        if let Some(ref mut rewards) = self.rewards.get_mut(sp) {
+            std::mem::take(&mut rewards.for_promotions)
+        } else {
+            Decimal::ZERO
+        }
+    }
+
+    pub fn into_mobile_reward_shares(
+        self,
+        reward_period: &'_ Range<DateTime<Utc>>,
+    ) -> impl Iterator<Item = proto::MobileRewardShare> + '_ {
+        self.rewards
+            .into_iter()
+            .map(|(service_provider_id, reward)| proto::MobileRewardShare {
+                start_period: reward_period.start.encode_timestamp(),
+                end_period: reward_period.end.encode_timestamp(),
+                reward: Some(ProtoReward::ServiceProviderReward(
+                    proto::ServiceProviderReward {
+                        service_provider_id,
+                        amount: (reward.for_promotions + reward.for_service_provider)
```

Comment:
It is a lil odd to read this and see the SP being rewarded the sum of for_promotions and for_service_provider.  Without being aware of the `take` above it is confusing....I wonder if there is a better way to approach this without things being as obfuscated

Reply:
I've been trying to think of a way to make this stage clearer as well. So far, nothing straight forward has come to mind. But please let me know if you have any ideas.